### PR TITLE
Fix for bug in PHP FPM through both ProxyPassMatch and SetHandler

### LINKED ISSUE #7719: owncloud 6.0.2 does not run under apache 2.4/php-fpm
my archlinux distribution just updated to apache 2.4. The recommended update for php was to use php-fpm and proxy-fcgi. 

Before the update owncloud 6.0.2 worked fine with apache 2.2 and php-module. After the update it does run: images/icons und css-styleheets are not loaded.

Similar problems are reported for ubuntu/debian.
https://forum.owncloud.org/viewtopic.php?f=26&t=19754

The following apache access log for the "home" /index.php and for the "files-directory" /index.php/apps/files  shows some very strange HTTP-GET requests:

GET /index.php//index.php/apps/files/css/upload.css

with double reference to index.php and "//"-separator. 

The problem might be due to the $_SERVER-environment provided by php-fpm. But owncloud should be able to run on this typical webserver-configuration.

-----                   clicking on the opencloud icon

88.67.235.79 - - [13/Mar/2014:17:40:16 +0100] "GET /index.php HTTP/1.1" 200 13344
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/file-upload.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/jquery.iframe-transport.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/jquery.fileupload.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/jquery-visibility.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/filelist.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/fileactions.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/files.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/keyboardshortcuts.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /index.php/core/js/config.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14352
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /remote.php/core.css?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /remote.php/core.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /index.php/apps/files/css/upload.css?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14800
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /index.php/apps/files/css/files.css?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14744
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /core/img/logo-wide.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /settings/img/admin.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /settings/img/personal.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /settings/img/users.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /core/img/actions/caret.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /core/img/breadcrumb.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /settings/img/help.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /core/img/actions/logout.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /core/img/places/files.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /apps/activity/img/activity.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /apps/documents/img/documents.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /core/img/places/picture.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /apps/calendar/img/calendar.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /apps/contacts/img/contacts.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /settings/img/apps.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /core/img/actions/download.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /core/img/actions/delete.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /core/img/filetypes/text.png HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /core/img/filetypes/folder.png HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /core/img/places/link.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /core/img/filetypes/application-pdf.png HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "POST /index.php/core/ajax/translations.php HTTP/1.1" 200 14856
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /index.php/core/routes.json HTTP/1.1" 200 14296
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /remote.p/core/img/noise.png?2013-12-17T15:48:20Z HTTP/1.1" 404 1865
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /cron.php HTTP/1.1" 200 20
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /remote.p/core/img/actions/search.svg?2013-12-17T15:48:20Z HTTP/1.1" 404 1865
88.67.235.79 - - [13/Mar/2014:17:40:47 +0100] "POST /index.php/core/ajax/translations.php HTTP/1.1" 200 14856
88.67.235.79 - - [13/Mar/2014:17:41:02 +0100] "POST /index.php/core/ajax/translations.php HTTP/1.1" 200 14856
88.67.235.79 - - [13/Mar/2014:17:41:07 +0100] "GET /core/img/actions/history.svg HTTP/1.1" 304 - 
88.67.235.79 - - [13/Mar/2014:17:41:07 +0100] "GET /core/img/actions/share.svg HTTP/1.1" 304 - 
88.67.235.79 - - [13/Mar/2014:17:41:07 +0100] "GET /core/img/actions/rename.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:41:02 +0100] "GET /undefined/templates/slideshow.html HTTP/1.1" 404 1865
88.67.235.79 - - [13/Mar/2014:17:41:07 +0100] "GET /index.php/core/ajax/share.php?fetch=getItemsSharedStatuses&itemType=file HTTP/1.1" 200 14464
88.67.235.79 - - [13/Mar/2014:17:41:07 +0100] "GET /remote.p/core/img/actions/delete.svg?2013-12-17T15:48:20Z HTTP/1.1" 404 1865
88.67.235.79 - - [13/Mar/2014:17:41:07 +0100] "GET /index.php/apps/files/ajax/scan.php?force=false&dir=&requesttoken=a12709e97e645b11f592 HTTP/1.1" 200 14744
88.67.235.79 - - [13/Mar/2014:17:41:10 +0100] "POST /index.php/core/ajax/translations.php HTTP/1.1" 200 14856

-----                   clicking on the files icon 

88.67.235.79 - - [13/Mar/2014:17:42:21 +0100] "GET /index.php/apps/files HTTP/1.1" 200 13960
88.67.235.79 - - [13/Mar/2014:17:42:25 +0100] "GET /index.php//index.php/apps/files/css/upload.css?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15416
88.67.235.79 - - [13/Mar/2014:17:42:26 +0100] "GET /index.php//index.php/core/js/config.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14968
88.67.235.79 - - [13/Mar/2014:17:42:26 +0100] "GET /index.php//remote.php/core.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14464
88.67.235.79 - - [13/Mar/2014:17:42:26 +0100] "GET /index.php//index.php/apps/files/css/files.css?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15360
88.67.235.79 - - [13/Mar/2014:17:42:27 +0100] "GET /index.php//remote.php/core.css?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14520
88.67.235.79 - - [13/Mar/2014:17:42:27 +0100] "GET /index.php//apps/files/js/file-upload.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15024
88.67.235.79 - - [13/Mar/2014:17:42:31 +0100] "GET /index.php//apps/files/js/jquery.iframe-transport.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15696
88.67.235.79 - - [13/Mar/2014:17:42:35 +0100] "GET /index.php//apps/files/js/jquery.fileupload.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15360
88.67.235.79 - - [13/Mar/2014:17:42:40 +0100] "GET /index.php//apps/files/js/jquery-visibility.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15360
88.67.235.79 - - [13/Mar/2014:17:42:45 +0100] "GET /index.php//apps/files/js/filelist.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14856
88.67.235.79 - - [13/Mar/2014:17:42:50 +0100] "GET /index.php//apps/files/js/fileactions.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15024
88.67.235.79 - - [13/Mar/2014:17:42:54 +0100] "GET /index.php//apps/files/js/files.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14688
88.67.235.79 - - [13/Mar/2014:17:42:59 +0100] "GET /index.php//apps/files/js/keyboardshortcuts.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15360
88.67.235.79 - - [13/Mar/2014:17:43:27 +0100] "GET /index.php//core/img/actions/caret.svg HTTP/1.1" 200 14912
88.67.235.79 - - [13/Mar/2014:17:43:27 +0100] "GET /index.php//settings/img/users.svg HTTP/1.1" 200 14688
88.67.235.79 - - [13/Mar/2014:17:43:27 +0100] "GET /index.php//settings/img/help.svg HTTP/1.1" 200 14632
88.67.235.79 - - [13/Mar/2014:17:43:27 +0100] "GET /index.php//core/img/logo-wide.svg HTTP/1.1" 200 14688

### PR DESCRIPTION
**Pull request by Caffe1neAdd1ct**
where the global variable SERVER SCRIPT_NAME isn't set correctly on through apache 2.4 to php-fpm.

fixes #7719

### CODE REVIEW

File: lib/private/AppFramework/Http/Request.php [hunk 1]
```diff
@@ -770,5 +770,18 @@ private function getOverwriteHost() {
...
+     * uri not just script name e.g /index.php/app/files instead of expected /index.php 
+     * @see https://bugs.php.net/bug.php?id=65641
+     * @param string $name
+     * @return string
+     */
+    private function getRawScriptName() {
+        if (strpos($this->server['SCRIPT_NAME'], '.php') !== false) {
+            $name = DIRECTORY_SEPARATOR . basename($this->server['SCRIPT_FILENAME']);
+        }
+        return $name;
```
Comment (DeepDiver1975) [inline lib/private/AppFramework/Http/Request.php]: $name will be undefined in case the if statement is not evaluated to true :-1:

**Comment by DeepDiver1975** [comment]
@Caffe1neAdd1ct thanks for your pull request - please reference bugs this is going to fix - THX

**Comment by RealRancor** [comment]
Workaround for a Bug in PHP which affects only Apache2 running with PHP-FPM: https://bugs.php.net/bug.php?id=65641

This might replace: https://github.com/owncloud/core/pull/17491

**Comment by Caffe1neAdd1ct** [comment]
Potential Bug References:

https://github.com/owncloud/core/issues/22411
https://github.com/owncloud/core/issues/21871
https://github.com/owncloud/core/issues/7719

File: lib/private/AppFramework/Http/Request.php (hunk 1 above)
Comment (Caffe1neAdd1ct) [inline lib/private/AppFramework/Http/Request.php]: Thanks @DeepDiver1975 I'll look at that now.

**Comment by wpoely86** [comment]
Backport request for RHEL 7: https://bugzilla.redhat.com/show_bug.cgi?id=1248077

**Comment by wpoely86** [comment]
@Caffe1neAdd1ct Can you show the apache config you're using for this? It doesn't seem to work for me with 9.0.4 on CentOS7.

**Comment by Caffe1neAdd1ct** [comment]
Manual tests i'm going to run through :

- [ ]  Test webapp with apache ProxyPassMatch FPM
- [ ]  Test webapp with apache SetHandler FPM
- [ ] Test webapp with apache mod_php
- [ ] Test webapp with nginx fpm setup

- [ ]  Test sync client with apache ProxyPassMatch FPM
- [ ]  Test sync client with apache SetHandler FPM
- [ ] Test sync client with apache mod_php
- [ ] Test sync client with nginx fpm setup

**Comment by Caffe1neAdd1ct** [comment]
@wpoely86 Apache config on my servers:

```
<VirtualHost *:80>

    ServerName owncloud.castle
    ServerAdmin support@castle
    DocumentRoot /usr/share/webapps/owncloud/

    SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1
    SetEnv HOME /usr/share/webapps/owncloud
    SetEnv HTTP_HOME /usr/share/webapps/owncloud
    
    DirectoryIndex index.php
    ProxyPassMatch ^/(.*\.php(/.*)?)$ fcgi://127.0.0.1:9001/usr/share/webapps/owncloud/$1

    <Directory "/usr/share/webapps/owncloud">
        AllowOverride All
        Require all granted
    </Directory>

    ErrorLog "/var/log/httpd/owncloud.error.log"
    CustomLog "/var/log/httpd/owncloud.access.log" combined

</VirtualHost>
```

**Comment by Caffe1neAdd1ct** [comment]
Looks like i have some tests to look at:

```

	Test Result (14 failures )
Test\AppFramework\Http\RequestTest::testGetPathInfoWithoutSetEnvGeneric.testGetPathInfoWithoutSetEnvGeneric with data set #0
Test\AppFramework\Http\RequestTest::testGetPathInfoWithoutSetEnvGeneric.testGetPathInfoWithoutSetEnvGeneric with data set #1
Test\AppFramework\Http\RequestTest::testGetPathInfoWithoutSetEnvGeneric.testGetPathInfoWithoutSetEnvGeneric with data set #2
Test\AppFramework\Http\RequestTest::testGetPathInfoWithoutSetEnvGeneric.testGetPathInfoWithoutSetEnvGeneric with data set #3
Test\AppFramework\Http\RequestTest::testGetPathInfoWithoutSetEnvGeneric.testGetPathInfoWithoutSetEnvGeneric with data set #4
Test\AppFramework\Http\RequestTest::testGetRawPathInfoWithoutSetEnvGeneric.testGetRawPathInfoWithoutSetEnvGeneric with data set #0
Test\AppFramework\Http\RequestTest::testGetRawPathInfoWithoutSetEnvGeneric.testGetRawPathInfoWithoutSetEnvGeneric with data set #1
Test\AppFramework\Http\RequestTest::testGetRawPathInfoWithoutSetEnvGeneric.testGetRawPathInfoWithoutSetEnvGeneric with data set #2
Test\AppFramework\Http\RequestTest::testGetRawPathInfoWithoutSetEnvGeneric.testGetRawPathInfoWithoutSetEnvGeneric with data set #3
Test\AppFramework\Http\RequestTest::testGetRawPathInfoWithoutSetEnvGeneric.testGetRawPathInfoWithoutSetEnvGeneric with data set #4
Show all failed tests >>>
```

I'll have a look into these plus making sure this change doesn't break other setups.

**Comment by DeepDiver1975** [comment]
still 15 failing unit tests - https://ci.owncloud.org/job/ownCloud/job/core/job/PR-25747/2/testReport/

**Comment by Caffe1neAdd1ct** [comment]
@DeepDiver1975 Working on them, running them again locally to find the issues through `./autotest.sh`

**Comment by Caffe1neAdd1ct** [comment]
Still an issue with the test env for phpunit putting /core/ on the front of requests, having another look

**Comment by Caffe1neAdd1ct** [comment]
The 210 failing tests look like they are due to the tests running with a document root of the folder above core, meaning owncloud is running in a sub folder of core.

Is this a supported configuration?

**Comment by Caffe1neAdd1ct** [comment]
Waiting for broken tests to be fixed in master, will pull the master branch when the tests pass and re-run the pull request branch.

**Comment by Caffe1neAdd1ct** [comment]
Pulled master changes which fix 4 broken tests (d2e20213f5f5ad9ea25bd46311157747ba7d4d22) jenkins build scheduled.

**Comment by Caffe1neAdd1ct** [comment]
Issue with integration tests setting up users and data through ocs cli.

**Comment by Caffe1neAdd1ct** [comment]
Issue located, script name fix is cutting off the /ocs/ subfolder from the script name. Writing tests and fixing issue hopefully today.

**Comment by Caffe1neAdd1ct** [comment]
Unit tests and integration tests passing on my local development environment, pushed changes and waiting on Jenkins PR build.

**Comment by Caffe1neAdd1ct** [comment]
PHPUnit tests have failed due to `An exception occured in driver: SQLSTATE[HY000] [1040] Too many connections`

This seems unrelated to the current changes and the MySQL PHPUnit tests are passing on my environment without any issues.

There were a lot of tests running when this PR branch was kicked off which may have caused this.

@DeepDiver1975 are you able to kick off this build again?

**Comment by DeepDiver1975** [comment]
@Caffe1neAdd1ct nice jobs - thx - can I ask you to squash the commits? THX

**Comment by Caffe1neAdd1ct** [comment]
@DeepDiver1975 i'm out at the moment, i think GitHub's UI automates this when merging in a pull request (https://help.github.com/articles/about-pull-request-merge-squashing/)

If not i can have a go Monday :+1:

**Comment by Caffe1neAdd1ct** [comment]
New dev env had a localhost git author set, i've amended the commit however cla is still showing the non existent contributor as required to sign the cla.

Happened when squashing the commits via git cli rather than through NetBeans.

**Pull request closed without merging**